### ISSUE TITLE
Caching hashcodes

### ISSUE THREAD

**Issue by rgabbard** [python-attrs/attrs#423]
We currently have an application which is a heavy user of attrs.  For certain operations, a majority of the application's time is being spent in the hash method of attrs objects.  The frozen attrs objects in question are complex and themselves contain other frozen attrs objects as fields (this goes several layers deep).  Being able to cache the hash code value by user request at upper levels of this object tree ((like `prehash` in [Immutables](https://immutables.github.io/immutable.html))) would lead to a significant performance improvement for us.

Issue https://github.com/python-attrs/attrs/issues/261 discussed some performance issues with hashing and touched on this.  However, this idea seemed to get lost in the following discussion, which is now closed, so it seems worth moving to its own issue.

On that issue @DRMacIver suggested the following:
* Add a `memoize_hash` flag to `attr.s` defaulting to `None`.
* When `memoize_hash` is `True`, store the hash on a field called `__hash`.
* If hash is `False` (or inferred `False` from a default) it is an error to pass non-`None` `memoize_hash`
* If `slots` and `memoize_hash` are both `True`, add a bonus `__hash` slot to the class to store it.
* If hash is `True`, `memoize_hash` defaults to `frozen` and not `slots`

**Comment by hynek** [comment]
I think I could live with that but I think `cache_hash` is a better term for people to understand?

Also it should be `__attrs_cached_hash__` or something so we stay within our own namespace.

There’s one thing that concerns me here tho: just because a class is frozen, we cannot make the assumption that it’s true for all of its attributes (although that would certainly be the technical correct behavior). Therefore I see no way around making it strictly opt-in.

**Comment by DRMacIver** [comment]
>  (although that would certainly be the technical correct behavior). 

Isn't the convention in Python that non-frozen things shouldn't be hashable at all? Is there widespread usage of mutable hashable objects?

(That being said, making it explicitly opt in seems perfectly reasonable)

**Comment by hynek** [comment]
It might be a convention but I’m quite positive that at least 90% of Python programmers do not know that.

**Comment by rgabbard** [comment]
I agree that strictly opt-in is safer.  

Here is my updated understanding of the plan:
```
* Add a `cache_hash` flag to attr.s defaulting to `False`.
* When `cache_hash` is `True`, store the hash on a field called `__attrs_cached_hash`.
* If hash is `False` (or inferred `False` from a default) it is an error to pass non-`False` `cache_hash`
* If `slots` and `cache_hash` are both `True`, add a bonus `__attrs_cached_hash` slot to the class to store it.
```

I have a long plane ride this weekend when I could try to implement this.  I am not familiar with the attrs codebase, but at a quick skim it looks like what needs doing is all in `_make.py`:

* adding the new flag to `attrs` (line 654, currently). 
* Validate no `cache_hash` without `hash` in `wrap` (line 787).
* update `_attrs_to_init_script` to create cache field (when applicable; line 1244).
* update `_make_hash` (line 865) to generate code which uses caching, if requested
* update docstring and other documentation

The cached hash could be either lazily or eagerly initialized.  Any preference?

**Comment by rgabbard** [comment]
Closed by #426
